rules/integer_overflow.go

```diff
@@ -0,0 +1,87 @@
+// (c) Copyright 2016 Hewlett Packard Enterprise Development LP
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rules
+
+import (
+	"github.com/securego/gosec"
+	"go/ast"
+)
+
+type integerOverflowCheck struct {
+	gosec.MetaData
+	calls gosec.CallList
+}
+
+func (i *integerOverflowCheck) ID() string {
+	return i.MetaData.ID
+}
+
+func (i *integerOverflowCheck) Match(node ast.Node, ctx *gosec.Context) (*gosec.Issue, error) {
+	varName := make(map[string]ast.Node)
+	var issue *gosec.Issue
+
+	// strconv.Atoi is a common function.
+	// To reduce false positives, This code detects code which is converted to int32/int16 only.
+	match := func(n ast.Node) bool {
+		switch n := n.(type) {
+		case *ast.AssignStmt:
+			for _, expr := range n.Rhs {
+				if callExpr, ok := expr.(*ast.CallExpr); ok && i.calls.ContainsCallExpr(callExpr, ctx, false) != nil {
+					if id, ok := n.Lhs[0].(*ast.Ident); ok && id.Name != "_" {
+						// Example:
+						//  v, _ := strconv.Atoi("1111")
+						// Add "v" to varName map
+						varName[id.Name] = n
+					}
+				}
+			}
+		case *ast.CallExpr:
+			if fun, ok := n.Fun.(*ast.Ident); ok {
+				if fun.Name == "int32" || fun.Name == "int16" {
+					if idt, ok := n.Args[0].(*ast.Ident); ok {
+						if n, ok := varName[idt.Name]; ok {
+							// Detect int32(v) and int16(v)
+							issue = gosec.NewIssue(ctx, n, i.ID(), i.What, i.Severity, i.Confidence)
+							return false
+						}
+					}
+				}
+			}
+		}
+		return true
+	}
+
+	ast.Inspect(node, match)
```

Comment:
I think that it is possible to leverage the `gosec.Context` which keeps the global state during analysis instead of calling again the `Inspect` on the entire tree.

Reply:
@ccojocar Hi! Thank you for your review.
I think your idea is more efficient.

I'll consider how to know whether `strconv.Atoi` and `int32()` are same function using gosec.Context.